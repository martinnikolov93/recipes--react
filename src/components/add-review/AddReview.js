import React, { useState, useContext } from 'react';
import { FaRegStar, FaStar } from 'react-icons/fa'
import { getCookie, isEmptyObject } from '../../utils/helpers';
import styles from './AddReview.module.css'
import RecipeReviewsWrapper from '../recipe-reviews-wrapper/RecipeReviewsWrapper';
import UserContext from '../../Context';
import { Link } from 'react-router-dom';
import { config } from '../../utils/constants'

const AddReview = (props) => {
    const [rating, setRating] = useState(0)
    const [comment, setComment] = useState('')
    const [ratingErrors, setRatingErrors] = useState(false)
    const [haveRated, setHaveRated] = useState(false)
    const [reviews, setReviews] = useState(props.reviews)

    const context = useContext(UserContext)

    const submitHandler = (e) => {
        e.preventDefault()

        let ratingErrors = validateRating(rating)
        setRatingErrors(ratingErrors)

        if (ratingErrors) {
            return
        }

        fetch(config.url.API_URL + '/review', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                rating,
                comment,
                recipeId: props.recipeId,
                token: getCookie('auth-token')
            })
        })
            .then((response) => response.json())
            .then((review) => {
                setReviews([...reviews, review])
                setHaveRated(true)
            })
            .catch((err) => console.log(err))
    }

    const changeHandlerRating = (event) => {
        let rating = event.currentTarget.dataset.rating
        setRating(rating)

        let ratingErrors = validateRating(rating)
        setRatingErrors(ratingErrors)
    }

    const changeHandlerComment = (event) => {
        setComment(event.target.value)
    }

    const validateRating = (rating) => {
        let ratingErrors = {}

        if (rating <= 0 || rating > 5) {
            ratingErrors['bad-rating'] = 'You forgot to rate it!'
        }

        if (isEmptyObject(ratingErrors)) {
            return false
        }

        return ratingErrors
    }

    return (
        <>
            {context.loggedIn
                ?
                <>
                    {!haveRated
                        ?
                        <>
                            <span data-rating="1" onClick={changeHandlerRating}>{rating > 0 ? <FaStar /> : <FaRegStar />}</span>
                            <span data-rating="2" onClick={changeHandlerRating}>{rating > 1 ? <FaStar /> : <FaRegStar />}</span>
                            <span data-rating="3" onClick={changeHandlerRating}>{rating > 2 ? <FaStar /> : <FaRegStar />}</span>
                            <span data-rating="4" onClick={changeHandlerRating}>{rating > 3 ? <FaStar /> : <FaRegStar />}</span>
                            <span data-rating="5" onClick={changeHandlerRating}>{rating > 4 ? <FaStar /> : <FaRegStar />}</span>
                            <div className="input-error-text">{ratingErrors ? Object.values(ratingErrors)[0] : null}</div>

                            <form onSubmit={submitHandler}>
                                <div>
                                    <div>Comment: (Optional)</div>
                                    <textarea className={styles['add-review-textarea']} onChange={changeHandlerComment}></textarea>
                                </div>
                                <div>
                                    <button>Add review</button>
                                </div>
                            </form>
                        </>
                        :
                        <h3>You review has been submited!</h3>
                    }
                </>
                :
                <div><Link to='/login'>Log in</Link> to write a review!</div>
            }
            <br />
            <h2>Reviews</h2>
            <hr className={styles['recipe-hr']} />
            <br /><br />
            <RecipeReviewsWrapper reviews={reviews} />
        </>
    )
}

export default AddReview