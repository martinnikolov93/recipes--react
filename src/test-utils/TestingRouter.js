import React from 'react';
import UserContext from '../Context';
import { BrowserRouter, Route } from 'react-router-dom'

const TestingRouter = ({user, children}) => {
    return (
        <BrowserRouter>
            <Route>
                <UserContext.Provider value={user}>
                    {children}
                </UserContext.Provider>
            </Route>
        </BrowserRouter>
    )
}

export default TestingRouter