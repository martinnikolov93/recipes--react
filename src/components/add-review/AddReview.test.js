import React from 'react';
import renderer from 'react-test-renderer'
import AddReview from './AddReview';
import TestingRouter from '../../test-utils/TestingRouter';

jest.mock('../../components/recipe-reviews-wrapper/RecipeReviewsWrapper', () => 'RecipeReviewsWrapper')

const userAuthDummy = {
    loggedIn: true,
    user: {
        email: 'test@test.com'
    }
}

const userUnauthDummy = {
    loggedIn: false,
}

describe('Add review component', () => {
    it('Should render authenticated add review component', () => {
        const tree = renderer.create(
            <TestingRouter user={userAuthDummy}>
                <AddReview />
            </TestingRouter>
        ).toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('Should render unauthenticated add review component', () => {
        const tree = renderer.create(
            <TestingRouter user={userUnauthDummy}>
                <AddReview />
            </TestingRouter>
        ).toJSON()

        expect(tree).toMatchSnapshot()
    })
})