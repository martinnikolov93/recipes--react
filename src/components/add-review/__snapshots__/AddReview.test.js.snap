// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add review component Should render authenticated add review component 1`] = `
Array [
  <span
    data-rating="1"
    onClick={[Function]}
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 576 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z"
      />
    </svg>
  </span>,
  <span
    data-rating="2"
    onClick={[Function]}
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 576 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z"
      />
    </svg>
  </span>,
  <span
    data-rating="3"
    onClick={[Function]}
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 576 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z"
      />
    </svg>
  </span>,
  <span
    data-rating="4"
    onClick={[Function]}
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 576 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z"
      />
    </svg>
  </span>,
  <span
    data-rating="5"
    onClick={[Function]}
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 576 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z"
      />
    </svg>
  </span>,
  <div
    className="input-error-text"
  />,
  <form
    onSubmit={[Function]}
  >
    <div>
      <div>
        Comment: (Optional)
      </div>
      <textarea
        className="add-review-textarea"
        onChange={[Function]}
      />
    </div>
    <div>
      <button>
        Add review
      </button>
    </div>
  </form>,
  <br />,
  <h2>
    Reviews
  </h2>,
  <hr
    className="recipe-hr"
  />,
  <br />,
  <br />,
  <RecipeReviewsWrapper />,
]
`;

exports[`Add review component Should render unauthenticated add review component 1`] = `
Array [
  <div>
    <a
      href="/login"
      onClick={[Function]}
    >
      Log in
    </a>
     to write a review!
  </div>,
  <br />,
  <h2>
    Reviews
  </h2>,
  <hr
    className="recipe-hr"
  />,
  <br />,
  <br />,
  <RecipeReviewsWrapper />,
]
`;
